File: server/BudgetBoard-Database/Models/Goal.cs
namespace BudgetBoard.Database.Models;

public class Goal
{
    public Guid ID { get; set; }
    public string Name { get; set; } = string.Empty;
    public DateTime? CompleteDate { get; set; }
    public decimal Amount { get; set; } = 0.0M;
    public decimal? InitialAmount { get; set; }
    public decimal? MonthlyContribution { get; set; }
    public ICollection<Account> Accounts { get; set; } = new List<Account>();
    public required Guid UserID { get; set; }
    public ApplicationUser? User { get; set; } = null!;
}

public class NewGoal
{
    public required string Name { get; set; }
    public DateTime? CompleteDate { get; set; }
    public required decimal Amount { get; set; }
    public decimal? InitialAmount { get; set; }
    public decimal? MonthlyContribution { get; set; }
    public required ICollection<string> AccountIds { get; set; }
}

File: client/src/types/goal.ts
import { IAccount } from './account';

export interface IGoalCreateRequest {
  name: string;
  completeDate: Date | null;
  amount: number;
  initialAmount: number | null;
  monthlyContribution: number | null;
  accountIds: string[];
}

export interface IGoalUpdateRequest {
  id: string;
  name: string;
  completeDate: Date | null;
  isCompleteDateEditable: boolean;
  amount: number;
  initialAmount: number | null;
  monthlyContribution: number | null;
  isMonthlyContributionEditable: boolean;
  accountIds: string[];
}

export interface IGoalResponse {
  id: string;
  name: string;
  completeDate: Date;
  isCompleteDateEditable: boolean;
  amount: number;
  initialAmount: number;
  monthlyContribution: number;
  isMonthlyContributionEditable: boolean;
  estimatedInterestRate: number | null;
  accounts: IAccount[];
  userID: string;
}

export enum GoalType {
  None = '',
  SaveGoal = 'saveGoal',
  PayGoal = 'payGoal',
}

export enum GoalCondition {
  TimedGoal = 'timedGoal',
  MonthlyGoal = 'monthlyGoal',
}

export enum GoalTarget {
  TargetBalanceGoal = 'targetBalanceGoal',
  TargetAmountGoal = 'targetAmountGoal',
}
